pkg/kubectl/cmd/set/set_volume.go

```diff
@@ -0,0 +1,980 @@
+		o.AddOpts.Type = ""
+	}
+	if len(o.AddOpts.ClaimSize) > 0 {
+		o.AddOpts.CreateClaim = true
+		if len(o.AddOpts.ClaimName) == 0 {
+			o.AddOpts.ClaimName = names.SimpleNameGenerator.GenerateName("pvc-")
+		}
+		q, err := kresource.ParseQuantity(o.AddOpts.ClaimSize)
+		if err != nil {
+			return fmt.Errorf("--claim-size is not valid: %v", err)
+		}
+		o.AddOpts.ClaimSize = q.String()
+	}
+	if len(o.AddOpts.DefaultMode) == 0 {
+		o.AddOpts.DefaultMode = "644"
+	}
+	switch strings.ToLower(o.AddOpts.ClaimMode) {
+	case strings.ToLower(string(v1.ReadOnlyMany)), "rom":
+		o.AddOpts.ClaimMode = string(v1.ReadOnlyMany)
+	case strings.ToLower(string(v1.ReadWriteOnce)), "rwo":
+		o.AddOpts.ClaimMode = string(v1.ReadWriteOnce)
+	case strings.ToLower(string(v1.ReadWriteMany)), "rwm":
+		o.AddOpts.ClaimMode = string(v1.ReadWriteMany)
+	case "":
+	default:
+		return errors.New("--claim-mode must be one of ReadWriteOnce (rwo), ReadWriteMany (rwm), or ReadOnlyMany (rom)")
+	}
+	return nil
+}
+
+func (o *VolumeOptions) RunVolume(args []string, f cmdutil.Factory) error {
+	builder := f.NewBuilder().
+		ContinueOnError().
+		NamespaceParam(o.DefaultNamespace).DefaultNamespace().
+		FilenameParam(o.EnforceNamespace, &o.FilenameOptions).
+		Flatten()
+
+	if !o.Local {
+		builder = builder.
+			LabelSelectorParam(o.Selector).
+			ResourceTypeOrNameArgs(o.All, args...).
+			Latest()
+	} else {
+		// if a --local flag was provided, and a resource was specified in the form
+		// <resource>/<name>, fail immediately as --local cannot query the api server
+		// for the specified resource.
+		if len(args) > 0 {
+			return resource.LocalResourceError
+		}
+
+		builder = builder.Local(f.ClientForMapping)
+	}
+
+	singleItemImplied := false
+	infos, err := builder.Do().IntoSingleItemImplied(&singleItemImplied).Infos()
+	if err != nil {
+		return err
+	}
+
+	if o.List {
+		listingErrors := o.printVolumes(infos)
+		if len(listingErrors) > 0 {
+			return cmdutil.ErrExit
+		}
+		return nil
+	}
+
+	updateInfos := []*resource.Info{}
+	// if a claim should be created, generate the info we'll add to the flow
+	if o.Add && o.AddOpts.CreateClaim {
+		claim := o.AddOpts.createClaim()
+		m, err := o.Mapper.RESTMapping(schema.GroupKind{Kind: "PersistentVolumeClaim"})
+		if err != nil {
+			return err
+		}
+		mapper := resource.ClientMapperFunc(o.RESTClientFactory)
+		client, err := mapper.ClientForMapping(m)
+		if err != nil {
+			return err
+		}
+		info := &resource.Info{
+			Mapping:   m,
+			Client:    client,
+			Namespace: o.DefaultNamespace,
+			Object:    claim,
+		}
+		infos = append(infos, info)
+		updateInfos = append(updateInfos, info)
+	}
+
+	patches, patchError := o.getVolumeUpdatePatches(infos, singleItemImplied)
+
+	if patchError != nil {
+		return patchError
+	}
+	if len(o.Output) > 0 || o.Local || o.DryRun {
+		allErrs := []error{}
+		for _, info := range infos {
+			if err := o.PrintObject(o.Cmd, o.Local, o.Mapper, info.Object, o.Out); err != nil {
```

Comment:
this maybe will occur error

Reply:
https://github.com/kubernetes/kubernetes/issues/55407

Reply:
and @smarterclayton doing refactor..

https://github.com/kubernetes/kubernetes/pull/55660#issuecomment-344172145